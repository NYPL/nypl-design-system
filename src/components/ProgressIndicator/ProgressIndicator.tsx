import React from "react";
import {
  Box,
  CircularProgress as ChakraCircularProgress,
  CircularProgressLabel as ChakraCircularProgressLabel,
  Progress as ChakraProgress,
  useMultiStyleConfig,
} from "@chakra-ui/react";

import {
  ProgressIndicatorSizes,
  ProgressIndicatorTypes,
} from "./ProgressIndicatorTypes";
import Label from "../Label/Label";

export interface ProgressIndicatorProps {
  /** Flag to render the component in a dark background. */
  darkMode?: boolean;
  /** ID that other components can cross reference for accessibility purposes. */
  id?: string;
  /** Whether the `ProgressIndicator` should be linear or circular. */
  indicatorType?: ProgressIndicatorTypes;
  /** Whether the progress animation should display because the `value` prop is
   * not known. When this is set to `true`, the `value` prop will be ignored. */
  isIndeterminate?: boolean;
  /** The text to display in an HTML `label` element. */
  labelText: string;
  /** Visually show or hide the label text. When set to `false`, then the label
   * text will be added to the parent component as its `aria-label` attribute. */
  showLabel?: boolean;
  /** The size of the linear or circular progress. */
  size?: ProgressIndicatorSizes;
  /** A number between 0 to 100 that defines the progress' value. */
  value?: number;
}

/**
 * A component that displays a progress status for any task that takes a long
 * time to complete or consists of multiple steps. Examples include downloading,
 * uploading, or processing.
 */
const ProgressIndicator: React.FC<ProgressIndicatorProps> = (
  props: ProgressIndicatorProps
) => {
  const {
    darkMode = false,
    id,
    indicatorType = ProgressIndicatorTypes.Linear,
    isIndeterminate = false,
    labelText,
    showLabel = true,
    size = ProgressIndicatorSizes.Default,
    value = 0,
  } = props;
  const styles = useMultiStyleConfig("ProgressIndicator", {
    darkMode,
    size,
  });
  let finalValue = value;
  if (finalValue < 0 || finalValue > 100) {
    console.warn(
      "ProgressIndicator: pass in a `value` between 0 and 100. Defaulting to 0."
    );
    finalValue = 0;
  }
  const progressProps = {
    id,
    // If the label is visually shown, associate it with the progress indicator.
    // Otherwise, the `aria-label` will be added.
    "aria-label": showLabel ? null : labelText,
    "aria-labelledby": showLabel ? `${id}-label` : null,
    // If `isIndeterminate` is true, then it overrides the `value` prop.
    isIndeterminate: isIndeterminate || null,
    value: isIndeterminate ? null : finalValue,
  };
  const progressComponent = (indicatorType) => {
    // Only display the percentage text for the default size, not in the
    // indeterminate state, and when `showLabel` is true.
    if (indicatorType === ProgressIndicatorTypes.Circular) {
      // For the small size, since the label won't be visible, we need to add
      // it to the parent component's `aria-label` attribute.
      if (size === ProgressIndicatorSizes.Small) {
        progressProps["aria-label"] = labelText;
      }
      return (
        <Box __css={styles.circularContainer}>
          <ChakraCircularProgress {...progressProps} sx={styles.circular}>
            {showLabel &&
              !isIndeterminate &&
              size !== ProgressIndicatorSizes.Small && (
                <ChakraCircularProgressLabel>
                  {finalValue}%
                </ChakraCircularProgressLabel>
              )}
          </ChakraCircularProgress>
          {showLabel && size !== ProgressIndicatorSizes.Small && (
            <Label id={`${id}-label`} htmlFor={id}>
              {labelText}
            </Label>
          )}
        </Box>
      );
    }
    // The Linear progress indicator is the default.
    return (
      <>
        {showLabel && (
          <Label id={`${id}-label`} htmlFor={id}>
            {labelText}
          </Label>
        )}
        <Box __css={styles.linearContainer}>
          <ChakraProgress {...progressProps} sx={styles.linear} />
          {showLabel && !isIndeterminate && (
            <Box __css={styles.linearPercentage}>{finalValue}%</Box>
          )}
        </Box>
      </>
    );
  };

  return <Box __css={styles}>{progressComponent(indicatorType)}</Box>;
};

export default ProgressIndicator;
