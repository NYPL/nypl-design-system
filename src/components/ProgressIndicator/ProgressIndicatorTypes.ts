export enum ProgressIndicatorSizes {
  Default = "default",
  Small = "small",
}
export enum ProgressIndicatorTypes {
  Circular = "circular",
  Linear = "linear",
}
