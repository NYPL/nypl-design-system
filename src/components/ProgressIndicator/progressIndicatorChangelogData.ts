/** This data is used to populate the ComponentChangelogTable component.
 *
 * date: string (when adding new entry during development, set value as "Prerelease")
 * version: string (when adding new entry during development, set value as "Prerelease")
 * type: "Bug Fix" | "New Feature" | "Update";
 * affects: array["Accessibility" | "Documentation" | "Functionality" | "Styles"];
 * notes: array (will render as a bulleted list, add one array element for each list element)
 */
import { ChangelogData } from "../../utils/ComponentChangelogTable";

export const changelogData: ChangelogData[] = [
  {
    date: "Prerelease",
    version: "Prerelease",
    type: "Update",
    affects: ["Documentation", "Functionality"],
    notes: [
      "Adds `labelPlacement` prop to allow for custom placement of a label for circular indicators.",
    ],
  },
  {
    date: "2024-03-14",
    version: "3.0.0",
    type: "Update",
    affects: ["Styles"],
    notes: ["Chakra 2.8 update."],
  },
];
