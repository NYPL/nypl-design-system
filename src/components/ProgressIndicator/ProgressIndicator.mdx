import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import * as ProgressIndicatorStories from "./ProgressIndicator.stories";
import ComponentChangelogTable from "../../utils/ComponentChangelogTable";
import { changelogData } from "./progressIndicatorChangelogData";
import Link from "../Link/Link";

<Meta of={ProgressIndicatorStories} />

# ProgressIndicator

| Component Version | DS Version   |
| ----------------- | ------------ |
| Added             | `0.25.4`     |
| Latest            | `Prerelease` |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#linear-type" target="_self">Linear Type</Link>}
- {<Link href="#circular-type" target="_self">Circular Type</Link>}
- {<Link href="#sizing" target="_self">Sizing</Link>}
- {<Link href="#labels" target="_self">Labels</Link>}
- {<Link href="#indeterminate-state" target="_self">Indeterminate State</Link>}
- {<Link href="#dark-mode" target="_self">Dark Mode</Link>}
- {<Link href="#get-input-values" target="_self">Get Input Values</Link>}
- {<Link href="#changelog" target="_self">Changelog</Link>}

## Overview

<Description of={ProgressIndicatorStories} />

## Component Props

<Canvas of={ProgressIndicatorStories.WithControls} />

<ArgTypes of={ProgressIndicatorStories.WithControls} />

## Accessibility

Chakra UI takes care of the internal accessibility attributes for this
component. Specifically, the `role` attribute is set to `progressbar` and the
`aria-valuenow` attribute is set to the value of the `value` prop, or the
percentage completion value. The min and max values are set to "0" and "100",
respectively, and are set with the `aria-valuemin` and `aria-valuemax`
attributes. This makes this component and its value(s) visible to screen
readers.

When `showLabel` is set to false, the `labelText` value will be set to the main
`<div>`'s `aria-label` attribute. This is the same div that contains the
`aria-valuemin`, `aria-valuemax`, and `aria-valuenow` attributes.

Resources:

- [DigitalA11y WAI-ARIA: ROLE=PROGRESSBAR](https://www.digitala11y.com/progressbar-role/)
- [MDN Using the progressbar role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_progressbar_role)
- [Chakra UI Progress](https://chakra-ui.com/docs/components/feedback/progress)
- [Chakra UI Circular Progress](https://chakra-ui.com/docs/components/feedback/circular-progress)

## Linear Type

Progress bars are preferred in vertically narrow areas such as tables, cards,
dialogs, etc.

<Canvas of={ProgressIndicatorStories.LinearType} />

## Circular Type

The circular progress type is preferred for large content areas and for
full-screen loading. Set the `indicatorType` prop to `"circular"` for this type.

<Canvas of={ProgressIndicatorStories.CircularType} />

## Sizing

The starting height for the progress bar is 4px on mobile and 8px for desktop.
The `size` prop can be used to optionally set the height to 4px for desktop
through the `ProgressIndicatorSizes.Small` value.

`size="small"`

<Canvas of={ProgressIndicatorStories.LinearSizes} />

The starting size for the circular progress is 48px and can be made smaller to
24px with the `size` prop. The small 24px size can be used for inline local
changes in content. Note that in the small size, the label text and the
percentage will not displayed.

`size="small"`

<Canvas of={ProgressIndicatorStories.CircularSizes} />

## Labels

The `labelText` value and the `value` percentage are displayed by default. They
can be hidden through the `showLabel` prop.

`showLabel={false}`

Accessibility Note: when `showLabel` is false, the `aria-label` prop is set in
the progress element to provide a description of the progress for screen
readers.

<Canvas of={ProgressIndicatorStories.Labels} />

For circular indicator variants, the `circularLabelPlacement` prop can be used to set the
placement of a label relative to the indicator. By default, the label will be
displayed below the indicator.

<Canvas of={ProgressIndicatorStories.CircularLabelPlacements} />

## Indeterminate State

When the `isIndeterminate` prop is set to true, the `value` prop is ignored and
the state is set to an animated indeterminate state. This is often used when the
exact value or progress of the task is unknown.

<Canvas of={ProgressIndicatorStories.IndeterminateState} />

## Dark Mode

The `darkMode` prop is deprecated and should not be used. This component comes
with built-in dark mode styles that display automatically if dark mode is turned
on.

## Get Input Values

In the following example, we are setting the `value` prop based on a timer that
increases the value every second by 10. Once it reaches 100, it resets to 0.
This is a very simple example using `React.useState` to manage the state but in
a real application the value would come from a data source or server.

<Source
  code={`
// Example code
function ProgressIndicatorExample() {
  const [value, setValue] = useState(0);
  useEffect(() => {
    const interval = setInterval(() => {
      setValue((value) => (value === 100 ? 0 : value + 10));
    }, 1000);
    return () => clearInterval(interval);
  }, []);
  return (
    <SimpleGrid columns={1} gap="grid.m">
      <ProgressIndicator
        id="example"
        labelText="Progress example"
        value={value}
      />
      <ProgressIndicator
        id="example-circular"
        indicatorType="circular"
        labelText="Progress example"
        value={value}
      />
    </SimpleGrid>
  );
}
`}
  language="jsx"
/>

<Canvas of={ProgressIndicatorStories.WorkingExample} />

## Changelog

<ComponentChangelogTable changelogData={changelogData} />
