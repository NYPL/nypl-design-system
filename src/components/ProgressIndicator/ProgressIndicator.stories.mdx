import { Box } from "@chakra-ui/react";
import {
  ArgsTable,
  Canvas,
  Description,
  Meta,
  Story,
} from "@storybook/addon-docs";
import { withDesign } from "storybook-addon-designs";

import ProgressIndicator from "./ProgressIndicator";
import SimpleGrid from "../Grid/SimpleGrid";
import { getCategory } from "../../utils/componentCategories";
import DSProvider from "../../theme/provider";

<Meta
  title={getCategory("ProgressIndicator")}
  component={ProgressIndicator}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=37638%3A23842",
    },
    jest: ["ProgressIndicator.test.tsx"],
  }}
  argTypes={{
    darkMode: {
      table: { defaultValue: { summary: false } },
    },
    id: { control: false },
    indicatorType: {
      table: { defaultValue: { summary: "linear" } },
    },
    isIndeterminate: {
      table: { defaultValue: { summary: false } },
    },
    showLabel: {
      table: { defaultValue: { summary: true } },
    },
    size: {
      table: { defaultValue: { summary: "default" } },
    },
    value: {
      table: { defaultValue: { summary: 0 } },
    },
  }}
/>

# ProgressIndicator

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.25.4`   |
| Latest            | `1.0.6`    |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [Linear Type](#linear-type)
- [Circular Type](#circular-type)
- [Sizing](#sizing)
- [Labels](#labels)
- [Indeterminate State](#indeterminate-state)
- [Dark Mode](#dark-mode)
- [Get Input Values](#get-input-values)

## Overview

<Description of={ProgressIndicator} />

## Component Props

<Canvas withToolbar>
  <Story
    name="ProgressIndicator with Controls"
    args={{
      darkMode: false,
      id: "progressIndicator-id",
      indicatorType: "linear",
      isIndeterminate: false,
      labelText: "Progress",
      showLabel: true,
      size: "default",
      value: 50,
    }}
  >
    {(args) => <ProgressIndicator {...args} />}
  </Story>
</Canvas>

<ArgsTable story="ProgressIndicator with Controls" />

## Accessibility

Chakra UI takes care of the internal accessibility attributes for this component.
Specifically, the `role` attribute is set to `progressbar` and the `aria-valuenow`
attribute is set to the value of the `value` prop, or the percentage completion
value. The min and max values are set to "0" and "100", respectively, and are set
with the `aria-valuemin` and `aria-valuemax` attributes. This makes this
component and its value(s) visible to screen readers.

When `showLabel` is set to false, the `labelText` value will be set to the
main `<div>`'s `aria-label` attribute. This is the same div that contains the
`aria-valuemin`, `aria-valuemax`, and `aria-valuenow` attributes.

Resources:

- [DigitalA11y WAI-ARIA: ROLE=PROGRESSBAR](https://www.digitala11y.com/progressbar-role/)
- [MDN Using the progressbar role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_progressbar_role)
- [Chakra UI Progress](https://chakra-ui.com/docs/components/feedback/progress)
- [Chakra UI Circular Progress](https://chakra-ui.com/docs/components/feedback/circular-progress)

## Linear Type

Progress bars are preferred in vertically narrow areas such as tables, cards,
dialogs, etc.

<Canvas>
  <Story
    name="Linear Type"
    argTypes={{
      indicatorType: { table: { disable: true } },
    }}
  >
    <ProgressIndicator
      id="linear"
      indicatorType="linear"
      labelText="Linear Progress Type"
      value={50}
    />
  </Story>
</Canvas>

## Circular Type

The circular progress type is preferred for large content areas and for
full-screen loading. Set the `indicatorType` prop to `"circular"` for this type.

<Canvas>
  <Story
    name="Circular Type"
    argTypes={{
      indicatorType: { table: { disable: true } },
    }}
  >
    <ProgressIndicator
      id="circular"
      indicatorType="circular"
      labelText="Linear Progress Type"
      value={50}
    />
  </Story>
</Canvas>

## Sizing

The starting height for the progress bar is 4px on mobile and 8px for desktop.
The `size` prop can be used to optionally set the height to 4px for desktop
through the `ProgressIndicatorSizes.Small` value.

`size="small"`

<Canvas>
  <DSProvider>
    <SimpleGrid columns={1} gap="grid.m">
      <ProgressIndicator
        id="default-size"
        labelText="Default 8px size"
        value={50}
      />
      <ProgressIndicator
        id="small-size"
        labelText="Small 4px size"
        size="small"
        value={50}
      />
    </SimpleGrid>
  </DSProvider>
</Canvas>

The starting size for the circular progress is 48px and can be made smaller to
24px with the `size` prop. The small 24px size can be used for inline local
changes in content. Note that in the small size, the label text and the
percentage will not displayed.

`size="small"`

<Canvas>
  <DSProvider>
    <SimpleGrid columns={1} gap="grid.m">
      <ProgressIndicator
        id="default-size"
        indicatorType="circular"
        labelText="Default 48px size"
        value={50}
      />
      <ProgressIndicator
        id="small-size"
        indicatorType="circular"
        labelText="Small 24px size"
        size="small"
        value={50}
      />
    </SimpleGrid>
  </DSProvider>
</Canvas>

## Labels

The `labelText` value and the `value` percentage are displayed by default. They
can be hidden through the `showLabel` prop.

`showLabel={false}`

Accessibility Note: when `showLabel` is false, the `aria-label` prop is set in
the progress element to provide a description of the progress for screen readers.

<Canvas>
  <DSProvider>
    <SimpleGrid columns={1} gap="grid.m">
      <ProgressIndicator
        id="label"
        labelText="This label will be added through aria-label"
        showLabel={false}
        value={50}
      />
      <ProgressIndicator
        id="label-hidden"
        indicatorType="circular"
        labelText="This label will be added through aria-label"
        showLabel={false}
        value={50}
      />
    </SimpleGrid>
  </DSProvider>
</Canvas>

## Indeterminate State

When the `isIndeterminate` prop is set to true, the `value` prop is ignored and
the state is set to an animated indeterminate state. This is often used when the
exact value or progress of the task is unknown.

<Canvas>
  <DSProvider>
    <SimpleGrid columns={1} gap="grid.m">
      <ProgressIndicator
        id="indeterminate"
        isIndeterminate
        labelText="Indeterminate state"
        value={50}
      />
      <ProgressIndicator
        id="indeterminate-circular"
        indicatorType="circular"
        isIndeterminate
        labelText="Indeterminate state"
        value={50}
      />
    </SimpleGrid>
  </DSProvider>
</Canvas>

## Dark Mode

When in "dark mode", set the `darkMode` prop to true. This will render the
progress bar and the label text in white for better contrast.

Note: the background is manually set to better showcase the `darkMode` prop.

<Canvas>
  <DSProvider>
    <Box bg="black" w="100%" h="200px" p="20px">
      <SimpleGrid columns={1} gap="grid.m">
        <ProgressIndicator
          darkMode
          id="darkmode"
          labelText="Dark Mode"
          value={50}
        />
        <ProgressIndicator
          darkMode
          id="darkmode-circular"
          indicatorType="circular"
          labelText="Dark Mode"
          value={50}
        />
      </SimpleGrid>
    </Box>
  </DSProvider>
</Canvas>

## Get Input Values

In the following example, we are setting the `value` prop based on a timer that
increases the value every second by 10. Once it reaches 100, it resets to 0.
This is a very simple example using `React.useState` to manage the state but in
a real application the value would come from a data source or server.

```tsx
// Example code
function ProgressIndicatorExample() {
  const [value, setValue] = React.useState(0);
  React.useEffect(() => {
    const interval = setInterval(() => {
      setValue((value) => (value === 100 ? 0 : value + 10));
    }, 1000);
    return () => clearInterval(interval);
  }, []);
  return (
    <SimpleGrid columns={1} gap="grid.m">
      <ProgressIndicator
        id="example"
        labelText="Progress example"
        value={value}
      />
      <ProgressIndicator
        id="example-circular"
        indicatorType="circular"
        labelText="Progress example"
        value={value}
      />
    </SimpleGrid>
  );
}
```

export function ProgressIndicatorExample() {
  const [value, setValue] = React.useState(0);
  React.useEffect(() => {
    const interval = setInterval(() => {
      setValue((value) => (value === 100 ? 0 : value + 10));
    }, 1000);
    return () => clearInterval(interval);
  }, []);
  return (
    <SimpleGrid columns={1} gap="grid.m">
      <ProgressIndicator
        id="example"
        labelText="Progress example"
        value={value}
      />
      <ProgressIndicator
        id="example-circular"
        indicatorType="circular"
        labelText="Progress example"
        value={value}
      />
    </SimpleGrid>
  );
}

<Canvas>
  <DSProvider>
    <ProgressIndicatorExample />
  </DSProvider>
</Canvas>
