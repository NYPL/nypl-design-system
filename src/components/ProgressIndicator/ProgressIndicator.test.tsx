import * as React from "react";
import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import renderer from "react-test-renderer";

import ProgressIndicator from "./ProgressIndicator";
import { ProgressIndicatorTypes } from "./ProgressIndicatorTypes";

describe("ProgressIndicator Accessibility", () => {
  it("passes axe accessibility for linear and circular types", async () => {
    const linearUtils = render(
      <ProgressIndicator labelText="Linear" value={50} />
    );
    const circularUtils = render(
      <ProgressIndicator
        labelText="Circular"
        value={50}
        indicatorType={ProgressIndicatorTypes.Circular}
      />
    );
    expect(await axe(linearUtils.container)).toHaveNoViolations();
    expect(await axe(circularUtils.container)).toHaveNoViolations();
  });

  it("passes axe accessibility for linear and circular types without labels", async () => {
    const linearUtils = render(
      <ProgressIndicator labelText="Linear" value={50} showLabel={false} />
    );
    const circularUtils = render(
      <ProgressIndicator
        labelText="Circular"
        value={50}
        indicatorType={ProgressIndicatorTypes.Circular}
        showLabel={false}
      />
    );
    expect(await axe(linearUtils.container)).toHaveNoViolations();
    expect(await axe(circularUtils.container)).toHaveNoViolations();
  });

  it("passes axe accessibility for linear and circular types for indeterminate state", async () => {
    const linearUtils = render(
      <ProgressIndicator labelText="Linear" value={50} isIndeterminate />
    );
    const circularUtils = render(
      <ProgressIndicator
        labelText="Circular"
        value={50}
        indicatorType={ProgressIndicatorTypes.Circular}
        isIndeterminate
      />
    );
    expect(await axe(linearUtils.container)).toHaveNoViolations();
    expect(await axe(circularUtils.container)).toHaveNoViolations();
  });

  it("passes axe accessibility for linear and circular types for dark mode", async () => {
    const linearUtils = render(
      <ProgressIndicator labelText="Linear" value={50} darkMode />
    );
    const circularUtils = render(
      <ProgressIndicator
        labelText="Circular"
        value={50}
        indicatorType={ProgressIndicatorTypes.Circular}
        darkMode
      />
    );
    expect(await axe(linearUtils.container)).toHaveNoViolations();
    expect(await axe(circularUtils.container)).toHaveNoViolations();
  });
});

describe("ProgressIndicator", () => {
  it("renders a label and a progressbar for the linear type", () => {
    render(<ProgressIndicator labelText="Linear" value={50} />);
    expect(screen.getByLabelText("Linear")).toBeInTheDocument();
    expect(screen.getByRole("progressbar")).toBeInTheDocument();
  });

  it("renders a label, a progressbar, and an svg for the circular type", () => {
    const { container } = render(
      <ProgressIndicator
        labelText="Circular"
        indicatorType={ProgressIndicatorTypes.Circular}
        value={50}
      />
    );
    expect(screen.getByLabelText("Circular")).toBeInTheDocument();
    expect(screen.getByRole("progressbar")).toBeInTheDocument();
    expect(container.querySelector("svg")).toBeInTheDocument();
  });

  it("renders the appropriate aria atttribute when the label is hidden", () => {
    render(
      <ProgressIndicator labelText="Linear" value={50} showLabel={false} />
    );
    expect(screen.getByLabelText("Linear")).toBeInTheDocument();
    expect(screen.getByRole("progressbar")).toHaveAttribute(
      "aria-label",
      "Linear"
    );
  });

  it("renders the value passed", () => {
    const { rerender } = render(
      <ProgressIndicator labelText="Linear" value={50} />
    );
    expect(screen.getByText("50%")).toBeInTheDocument();

    rerender(<ProgressIndicator labelText="Linear" value={89} />);
    expect(screen.getByText("89%")).toBeInTheDocument();

    rerender(<ProgressIndicator labelText="Linear" value={4} />);
    expect(screen.getByText("4%")).toBeInTheDocument();
  });

  it("logs a warning if a value less than 0 is passed", () => {
    const warn = jest.spyOn(console, "warn");
    render(<ProgressIndicator labelText="Linear" value={-20} />);

    expect(warn).toHaveBeenCalledWith(
      "ProgressIndicator: pass in a `value` between 0 and 100. Defaulting to 0."
    );
  });

  it("logs a warning if a value more than 100 is passed", () => {
    const warn = jest.spyOn(console, "warn");
    render(<ProgressIndicator labelText="Linear" value={150} />);

    expect(warn).toHaveBeenCalledWith(
      "ProgressIndicator: pass in a `value` between 0 and 100. Defaulting to 0."
    );
  });

  it("Renders the UI snapshot correctly", () => {
    const linearBasic = renderer
      .create(<ProgressIndicator labelText="Linear" value={50} />)
      .toJSON();
    const circularBasic = renderer
      .create(
        <ProgressIndicator
          labelText="Circular"
          value={50}
          indicatorType={ProgressIndicatorTypes.Circular}
        />
      )
      .toJSON();
    const linearNoLabel = renderer
      .create(
        <ProgressIndicator labelText="Linear" value={50} showLabel={false} />
      )
      .toJSON();
    const circularNoLabel = renderer
      .create(
        <ProgressIndicator
          labelText="Circular"
          value={50}
          indicatorType={ProgressIndicatorTypes.Circular}
          showLabel={false}
        />
      )
      .toJSON();
    const linearIndeterminate = renderer
      .create(
        <ProgressIndicator labelText="Linear" value={50} isIndeterminate />
      )
      .toJSON();
    const circularIndeterminate = renderer
      .create(
        <ProgressIndicator
          labelText="Circular"
          value={50}
          indicatorType={ProgressIndicatorTypes.Circular}
          isIndeterminate
        />
      )
      .toJSON();
    const linearDarkMode = renderer
      .create(<ProgressIndicator labelText="Linear" value={50} darkMode />)
      .toJSON();
    const circularDarkMode = renderer
      .create(
        <ProgressIndicator
          labelText="Circular"
          value={50}
          indicatorType={ProgressIndicatorTypes.Circular}
          darkMode
        />
      )
      .toJSON();

    expect(linearBasic).toMatchSnapshot();
    expect(circularBasic).toMatchSnapshot();
    expect(linearNoLabel).toMatchSnapshot();
    expect(circularNoLabel).toMatchSnapshot();
    expect(linearIndeterminate).toMatchSnapshot();
    expect(circularIndeterminate).toMatchSnapshot();
    expect(linearDarkMode).toMatchSnapshot();
    expect(circularDarkMode).toMatchSnapshot();
  });
});
