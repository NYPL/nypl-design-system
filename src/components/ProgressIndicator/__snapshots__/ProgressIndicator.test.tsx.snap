// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressIndicator Renders the UI snapshot correctly 1`] = `
<div
  className="css-1xdhyk6"
>
  <label
    className="css-1xdhyk6"
    htmlFor="linearBasic"
    id="linearBasic-label"
  >
    Linear
  </label>
  <div
    className="css-0"
  >
    <div
      className="css-kdwx3d"
      id="linearBasic"
    >
      <div
        aria-label={null}
        aria-labelledby="linearBasic-label"
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={50}
        className="css-1jy0d03"
        role="progressbar"
        style={
          Object {
            "width": "50%",
          }
        }
      />
    </div>
    <div
      className="css-0"
    >
      50
      %
    </div>
  </div>
</div>
`;

exports[`ProgressIndicator Renders the UI snapshot correctly 2`] = `
<div
  className="css-1xdhyk6"
>
  <div
    className="css-0"
  >
    <div
      aria-label={null}
      aria-labelledby="circularBasic-label"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={50}
      className="chakra-progress css-120wkjd"
      id="circularBasic"
      role="progressbar"
    >
      <svg
        className="css-kn46u7"
        viewBox="0 0 100 100"
      >
        <circle
          className="chakra-progress__track css-1bhxzgy"
          cx={50}
          cy={50}
          r={42}
          strokeWidth="10px"
        />
        <circle
          className="chakra-progress__indicator css-5hfhx0"
          cx={50}
          cy={50}
          r={42}
          strokeDasharray="132 132"
          strokeDashoffset={66}
          strokeWidth="10px"
        />
      </svg>
      <div
        className="css-f1j64i"
      >
        50
        %
      </div>
    </div>
    <label
      className="css-1xdhyk6"
      htmlFor="circularBasic"
      id="circularBasic-label"
    >
      Circular
    </label>
  </div>
</div>
`;

exports[`ProgressIndicator Renders the UI snapshot correctly 3`] = `
<div
  className="css-1xdhyk6"
>
  <div
    className="css-0"
  >
    <div
      className="css-kdwx3d"
      id="linearNoLabel"
    >
      <div
        aria-label="Linear"
        aria-labelledby={null}
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={50}
        className="css-1jy0d03"
        role="progressbar"
        style={
          Object {
            "width": "50%",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`ProgressIndicator Renders the UI snapshot correctly 4`] = `
<div
  className="css-1xdhyk6"
>
  <div
    className="css-0"
  >
    <div
      aria-label="Circular"
      aria-labelledby={null}
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={50}
      className="chakra-progress css-120wkjd"
      id="circularNoLabel"
      role="progressbar"
    >
      <svg
        className="css-kn46u7"
        viewBox="0 0 100 100"
      >
        <circle
          className="chakra-progress__track css-1bhxzgy"
          cx={50}
          cy={50}
          r={42}
          strokeWidth="10px"
        />
        <circle
          className="chakra-progress__indicator css-5hfhx0"
          cx={50}
          cy={50}
          r={42}
          strokeDasharray="132 132"
          strokeDashoffset={66}
          strokeWidth="10px"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`ProgressIndicator Renders the UI snapshot correctly 5`] = `
<div
  className="css-1xdhyk6"
>
  <label
    className="css-1xdhyk6"
    htmlFor="linearIndeterminate"
    id="linearIndeterminate-label"
  >
    Linear
  </label>
  <div
    className="css-0"
  >
    <div
      className="css-kdwx3d"
      id="linearIndeterminate"
    >
      <div
        aria-label={null}
        aria-labelledby="linearIndeterminate-label"
        aria-valuemax={100}
        aria-valuemin={0}
        className="css-1demtxk"
        data-indeterminate=""
        role="progressbar"
        style={
          Object {
            "width": "0%",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`ProgressIndicator Renders the UI snapshot correctly 6`] = `
<div
  className="css-1xdhyk6"
>
  <div
    className="css-0"
  >
    <div
      aria-label={null}
      aria-labelledby="circularIndeterminate-label"
      aria-valuemax={100}
      aria-valuemin={0}
      className="chakra-progress css-120wkjd"
      data-indeterminate=""
      id="circularIndeterminate"
      role="progressbar"
    >
      <svg
        className="css-jf70f3"
        viewBox="0 0 100 100"
      >
        <circle
          className="chakra-progress__track css-1bhxzgy"
          cx={50}
          cy={50}
          r={42}
          strokeWidth="10px"
        />
        <circle
          className="chakra-progress__indicator css-tueqm1"
          cx={50}
          cy={50}
          r={42}
          strokeWidth="10px"
        />
      </svg>
    </div>
    <label
      className="css-1xdhyk6"
      htmlFor="circularIndeterminate"
      id="circularIndeterminate-label"
    >
      Circular
    </label>
  </div>
</div>
`;

exports[`ProgressIndicator Renders the UI snapshot correctly 7`] = `
<div
  className="css-1xdhyk6"
>
  <label
    className="css-1xdhyk6"
    htmlFor="linearDarkMode"
    id="linearDarkMode-label"
  >
    Linear
  </label>
  <div
    className="css-0"
  >
    <div
      className="css-kdwx3d"
      id="linearDarkMode"
    >
      <div
        aria-label={null}
        aria-labelledby="linearDarkMode-label"
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={50}
        className="css-1jy0d03"
        role="progressbar"
        style={
          Object {
            "width": "50%",
          }
        }
      />
    </div>
    <div
      className="css-0"
    >
      50
      %
    </div>
  </div>
</div>
`;

exports[`ProgressIndicator Renders the UI snapshot correctly 8`] = `
<div
  className="css-1xdhyk6"
>
  <div
    className="css-0"
  >
    <div
      aria-label={null}
      aria-labelledby="circularDarkMode-label"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={50}
      className="chakra-progress css-120wkjd"
      id="circularDarkMode"
      role="progressbar"
    >
      <svg
        className="css-kn46u7"
        viewBox="0 0 100 100"
      >
        <circle
          className="chakra-progress__track css-1bhxzgy"
          cx={50}
          cy={50}
          r={42}
          strokeWidth="10px"
        />
        <circle
          className="chakra-progress__indicator css-5hfhx0"
          cx={50}
          cy={50}
          r={42}
          strokeDasharray="132 132"
          strokeDashoffset={66}
          strokeWidth="10px"
        />
      </svg>
      <div
        className="css-f1j64i"
      >
        50
        %
      </div>
    </div>
    <label
      className="css-1xdhyk6"
      htmlFor="circularDarkMode"
      id="circularDarkMode-label"
    >
      Circular
    </label>
  </div>
</div>
`;
